feat: add JSR protocol support for package specifiers

eser opened the pull request:
**Motivation**

Given that JSR is an MIT-licensed open-source solution and is already supported by other package managers such as Volt and pnpm, I believe npm should also adopt this feature to maintain consistency within the JavaScript ecosystem and provide a seamless user experience.

**Current Behavior**

Before this change, installing JSR packages is possible by adding the following to the `package.json`:

```json
{
  "dependencies": {
    "@scope/package": "npm:@jsr/scope__package@1.0.0"
  }
}
```

This approach is rather messy and relies on the [npm compatibility registry API](https://jsr.io/docs/api#npm-compatibility-registry-api) to download packages from the JSR registry.

**Proposed Solution**

The proposed solution would be to handle installing packages from JSR natively. Installing a package from JSR would be simplified to:

```bash
npm i jsr:@scope/package
```

Which then installs a JSR package into the `package.json` as follows:

```json
{
  "dependencies": {
    "@scope/package": "jsr:@scope/package@1.0.0"
  }
}
```

**Reference**

This implementation is based on the approach discussed in https://github.com/pnpm/pnpm/issues/8941.

eser committed:
feat: add JSR protocol support for package specifiers

wraithgar commented:
It seems that parsing a jsr: protocol entry is identical to parsing a scoped package range after you've sliced off the `jsr:` part.  If that's so we don't need to re-implement a scoped package parser.  Much like how  `fromAlias` works, it parses the reference without `npm:` and then tweaks the response object.  We can do something similar, but probably skip straight to the `fromRegistry` part.

eser committed:
chore: reusing existing package parser routines

ljharb commented on lib/npa.js:
```diff
@@ -453,6 +462,46 @@ function fromAlias (res, where) {
 
+function fromJsr (res, where) {
+  // Remove 'jsr:' prefix
+  const jsrSpec = res.rawSpec.substr(4)
```
```suggestion
  const jsrSpec = res.rawSpec.slice(4)
```
substr is legacy

eser committed:
fix: enforce 'jsr:' prefix for import specifiers

Also replaces String.prototype.substr with #slice as the API is considered legacy.

eser committed:
test: updated tests to cover JSR case sensitivity criteria.

ljharb commented:
It empirically caused a huge problem when npm first added `^` and packages started using it, so much so that some packages (express) waited over a decade before depending on it. The presence of new protocols in packages like jackspeak has caused me a ton of issues in older engines.

This new specifier, if published in packages, will *absolutely* cause ecosystem breakage. If there's a way to only support it in applications (ie, only the top-level package.json, not transitive deps) that would be ideal.

anonrig commented:
Any updates on this PR? Anything I can do to help?

owlstronaut commented on lib/npa.js:
```diff
@@ -98,6 +105,8 @@ function resolve (name, spec, where, arg) {
   } else if (isAliasSpec(spec)) {
     return fromAlias(res, where)
+  } else if (isJsrSpec(spec)) {
+    return fromJsr(res, where)
```
What happens when `resolve` is called with a preexisting name?

owlstronaut commented on lib/npa.js:
```diff
@@ -453,6 +462,46 @@ function fromAlias (res, where) {
+    throw new Error(`JSR packages must be scoped (e.g., jsr:@scope/name): ${res.raw}`)
+  }
+
+  const subSpec = npa(jsrSpec, where)
```
This accepts nested protocols without validation

owlstronaut commented on lib/npa.js:
```diff
@@ -453,6 +462,46 @@ function fromAlias (res, where) {
+  // Set the transformed name and copy properties from subSpec
+  res.setName(transformedName)
+  res.registry = true
+  res.type = subSpec.type
```
How will consumers identify JSR deps for filtering?
